[feature] 카드 기능 추가, 체크리스트 및 체크 기능 개발 후 연관관계 매핑

sxi8022 opened the pull request:
## 관련 Issue

<!-- 해당 Pull Request와 관련된 Issue를 적습니다. -->

#39 

## 변경 사항

 1. 카드 단일 조회 기능 추가
   - 단일 조회 시, 카드정보와 체크리스트 목록이 나오도록 처리 
 2. 카드 엔티티에 created_at, modified_at 컬럼 추가 (튜터님 피드백)   
 3. 체크리스트 CRUD 기능 적용  
 4. Checks CRUD 기능 적용
 5. 테스트 코드 작성 
 - 범위 : 카드, 체크리스트, 체크 CRUD 관련 테스트 케이스 작성 
 (테스트 코드를 활용한 테스트는 별도의 이슈 발행하고 포스트맨을 활용한 테스트로 대체) 
 6. 댓글 엔티티의 닉네임 주석 처리 해제 
 7. 카드 삭제 시 별도로 체크 삭제, 체크리스트 삭제를 해야 삭제가 가능하도록 로직 구현 

## Todo List (별도의 이슈 생성 예정)
1. 체크박스와 체크리스트도 테스트코드로 테스트 예정 
   -> mock 테스트코드를 작성하였음
   -> 카드 crud는 테스트코드로 테스트 완료 , 체크박스와 체크리스트도 테스트코드로 테스트 예정   
2. 카드컨트롤러에서 댓글 조회 기능 추가 예정 (Pageable 기능 추가) 

## Check List


- [x] 포스트맨으로 체크해 보았나요?
- [x] 테스트 코드를 작성하셨나요? 

- [x] 일부 체크리스트는 추가로 테스트 필요함 별도의 이슈생성 예정

## 트러블 슈팅

- 테이블 명을 check 라고 했더니 테이블이 생성되지 않음 

원인 분석

> mysql은 check라는 단어를 예약어로 받아들인다. 그래서 checks라고 테이블명 수정 

처리
> 클래스명 전부 Check -> Checks 로 변경 

## 테스트 코드 사용시 주의사항 
- user라는 계정으로 회원가입 후에 진행할 것 

## Postman 테스트사진 

![230810_체크등록](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/6cab4ce9-450f-4dae-8ade-e659dc4486bf)

![230810_체크리스트등록](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/e8452654-a5f6-4411-9e2f-3ce761b91759)

![230810_체크리스트삭제](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/f150f495-8828-45a1-863e-fa987b94e3a6)
![230810_체크리스트상세조회](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/4da50217-ef99-4b62-a794-c464e08b53ce)
![230810_체크리스트수정](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/ff892c9b-d0b3-4ec1-a917-95f18c953c26)
![230810_체크삭제](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/44abd987-29a0-46c7-ad47-af173eb438d7)
![230810_체크수정](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/41d09ad5-4bdb-4528-993c-e0b48ff27b8a)
![230810_체크이름수정](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/8e04dedf-7dba-4943-9624-c88ca949eec1)
![230810_카드삭제시자식테이블부터삭제되도록처리](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/c5801db7-2963-44da-9cad-a3be3ba0d1e9)
![230810_카드상세화면에서 체크목록나오게처리](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/aae7e1d2-ddc5-4a36-ae7a-92698782101b)

sxi8022 committed:
[update] 카드 상세조회 및 테스트코드 작성

sxi8022 committed:
[update] 카드 기능 추가 및 연관관계 매핑
- 파일첨부 테스트를 위해 리소스 하위 경로 파일 이그노어 설정 추가

sxi8022 committed:
[update] 카드 기능 추가 및 연관관계 매핑
 1. 카드 단일 조회 기능 추가
   - 단일 조회 시, 카드정보와 댓글 목록 표시 필요
   -  체크리스트 목록이 나오도록 처리
 2. 카드 엔티티에 created_at, modified_at 컬럼 추가
 3. 체크리스트 CRUD 기능 적용
 4. Checks CRUD 기능 적용
 5. 테스트 코드 작성
 6. 댓글 엔티티의 닉네임 주석 처리 해제
 7. 카드 삭제 시 별도로 체크 삭제, 체크리스트 삭제를 해야 삭제가 가능하도록 로직 구현

proLmpa commented on src/main/java/com/example/itwassummer/card/service/CardService.java:
```diff
@@ -39,31 +48,31 @@ CardResponseDto update(Long cardId, CardRequestDto requestDto, List<MultipartFil
   CardResponseDto changeDueDate(Long cardId, String dueDate);
 
   /**
    * 카드 마감일 수정
```
카드 이동이 맞는 거 같네용

proLmpa replied:
카드 마감일 수정이 중복으로 존재해서 말씀드린 겁니다!

sxi8022 replied:
처리했습니다

proLmpa commented on src/main/java/com/example/itwassummer/card/service/CardServiceImpl.java:
```diff
@@ -36,10 +40,50 @@ public class CardServiceImpl implements CardService {
+    CardViewResponseDto responseDto = null;
+
+    if (card != null) {
+      responseDto = CardViewResponseDto
```
이거 builder 방식이 내부에 어떤 인자가 있는지 보여주는 거다 보니까 협업자가 한눈에 보고 이해할 수 있겠지만

여기는 Service 구현 부분이라 그냥 평범하게 responseDto = CardViewResponseDto(card); 로 하고 데이터는 CardViewResponseDto.class 에서 처리하는 게 좋겠습니다.

이 메서드 외에 동일한 방식으로 구현한 메서드가 있다면 위에 말씀드린 방법 적용을 고려해주세요

sxi8022 replied:
서비스에서 빌더패턴 대신 생성자 패턴을 사용하여 dto 객체 출력

proLmpa commented on src/main/java/com/example/itwassummer/checklist/service/CheckListServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+  public CheckListResponseDto getCheckList(Long listId) {
+    CheckList checkList = this.findCheckList(listId);
+    CheckListResponseDto responseDto = null;
+    if (checkList != null) {
```
굳이 null 확인 조건문을 쓰실 필요는 없으실 게 이미 위에 작성한 findCheckList 함수를 통해 만약 checkList 가 없으면 CustomException을 통해 에러 메시지를 출력하고 종료하게 되도록 RestControllerAdvice 를 정의 했기 때문에 이 메서드 포함 다른 메서드에 작성된 동일한 null 확인 조건문은 삭제하시는 게 좋겠습니다.

sxi8022 replied:
처리완료

proLmpa commented on src/test/java/com/example/itwassummer/card/controller/CardControllerTest.java:
```diff
@@ -35,64 +57,153 @@ public class CardControllerTest {
     LocalDateTime now = LocalDateTime.now();
 
-    // when
     String body = mapper.writeValueAsString(
```
테스트 코드처럼 개발자가 동작을 이해하기 위한 테스트 코드에선 작성하신 builder 방식이 잘 어울리는 거 같아서 테스트 코드에서 builder 방식은 그대로 두셔도 될 거 같습니다.

sxi8022 replied:
서비스에서 빌더패턴 대신 생성자 패턴을 사용하여 dto 객체 출력

proLmpa reviewed (commented):
피드백 입니다 :D

sxi8022 committed:
[update] 카드 기능 추가 및 연관관계 매핑
- 카드 마감일 수정 주석이 중복으로 들어가는 문제수정
- 카드 테스트 케이스 소스코드 수정
- 카드 순서 이동 시 원래 순서를 가지고있는 행의 정보가 수정되지 않는 현상 수정
- 카드, 체크리스트, 체크 서비스 파일들은 builder 방식 소스코드를 생성자 선언 방식으로 수정
- 불필요 null 체크 제거

proLmpa reviewed (approved):
확인했습니다~

ayboori reviewed (commented):
확인했습니다! 수고하셨습니다 :)
